### PR TITLE
[api] [database] New, migrate to new FAB API side by side

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [X] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Refactor DatabaseAsync API, using FAB's new `ModelRestApi`, aim to:

- Use Rison for parameters
- Coherent API JSON responses
- Coherent namespace
- Swagger UI

WIP, pending:
- Check, pre_add, pre_update, pre_delete
 
NOTE:
Add `FAB_API_SWAGGER_UI = True` to enable swagger UI, accessible on: http://localhost:8080/swaggerview/v1
Yet superset init does not create permissions for it

### CODE REVIEW

**Comment by mistercrunch** [comment]
As you touch this, it would be good to start moving things off of `core.py` and onto new smaller modules. Personally wouldn't mind having one entity per module `views/model_views/database.py`

**Comment by dpgaspar** [comment]
Made a major change on database and views module reorganization.

This is still very much a work in progress, @mistercrunch take a look and see if you agree.
I would personally end with the `views` module, and would create modules by function, using the following structure:

```
superset/
    database/
       __init__.py
       views.py
       api.py
       forms.py
       models.py (putting the models here is organized, but is highly debatable)
``` 

What I have done on this PR is more conservative:
```
superset/
    views/
        database/
             __init__.py
            views.py
            api.py
            forms.py
```

But `superset.views.database.views` is a bit redundant on `views` right?

File: superset/views/database/forms.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+from wtforms.validators import DataRequired, Length, NumberRange, Optional
+
+from superset import app, db, security_manager
+from superset.forms import CommaSeparatedListField, filter_not_empty_values
+from superset.models import core as models
+
+config = app.config
+
+
+class CsvToDatabaseForm(DynamicForm):
```
Comment (mistercrunch) [inline superset/views/database/forms.py]: I'm not sure if that belongs under here in `views/database/...`

**Review by mistercrunch — COMMENTED** [review]
In terms of folder structure I think that's the right approach. `superset.views.database.views` is fine by me. Maybe in the future we may have a more model centric approach, but for now the goal is probably just to break down `views/` further.

You probably just not have gotten there yet, but I'd also assume we'd move some routes on the javascript side to go from `databaseasync` to `api/`

File: superset/config.py [hunk 2]
```diff
@@ -645,3 +645,6 @@ class CeleryConfig(object):
         )
 except ImportError:
     pass
+
+SQLALCHEMY_DATABASE_URI = 'postgresql://postgres:postgres@localhost/superset'
```
Comment (mistercrunch) [inline superset/config.py]: I think this should be in your local `superset_config.py`

Comment (dpgaspar) [reply superset/config.py]: yes, my bad

File: superset/assets/src/components/TableSelector.jsx [hunk 3]
```diff
@@ -218,10 +218,11 @@ export default class TableSelector extends React.PureComponent {
...
         dataEndpoint={
-          '/databaseasync/api/' +
-          'read?_flt_0_expose_in_sqllab=1&' +
-          '_oc_DatabaseAsync=database_name&' +
-          '_od_DatabaseAsync=asc'
+          '/api/v1/databaseasync/?q=' +
+          '(keys:!(none),' +
+          'columns:!(id,database_name,backend),' +
+          'filters:!((col:expose_in_sqllab,opr:eq,value:true)),' +
+          'order_columns:database_name,order_direction:asc)'
```
Comment (dpgaspar) [inline superset/assets/src/components/TableSelector.jsx]: Here we only ask for `ids` and `result` keys, and exactly the columns we need

File: superset/assets/src/SqlLab/components/QuerySearch.jsx [hunk 4]
```diff
@@ -207,7 +207,7 @@ class QuerySearch extends React.PureComponent {
           <div className="col-sm-2">
             <AsyncSelect
               onChange={this.onChange}
-              dataEndpoint="/databaseasync/api/read?_flt_0_expose_in_sqllab=1"
+              dataEndpoint="/api/v1/databaseasync/?q=(filters:!((col:expose_in_sqllab,opr:eq,value:true))"
```
Comment (dpgaspar) [inline superset/assets/src/SqlLab/components/QuerySearch.jsx]: @mistercrunch can you help, where is this used? I can make this call more efficient also, take a look at the other on. I mean we can only retrieve the keys and columns we want

**Comment by dpgaspar** [comment]
Still need to change the tests for tests.security_tests.RolePermissionTests

**Comment by dpgaspar** [comment]
Starting to look good!

**Comment by dpgaspar** [comment]
I have not introduced `pre_add`, `pre_update` and `pre_delete` hooks on the new API, I'm assuming it is strictly used for read operation (that's the case on superset by itself). Is this a correct assumption?

**Comment by mistercrunch** [comment]
Currently that is the case (only used for read), but moving forward we'll move away from the CRUD in FAB and aim at the the Rest API, I'd hook those up now as this move pattern is likely to be repeated too. You could implement as a mixin that lives in the new module and is reused on ModelView and RestModelView.

**Review by mistercrunch — APPROVED** [review]
LGTM

**Pull request merged**